gtest_parallel_mocks.py

```diff
@@ -0,0 +1,184 @@
+
+  def record_test_time(self, test_binary, test_name, last_execution_time):
+    test_id = (test_binary, test_name)
+    self.last_execution_times[test_id].append(last_execution_time)
+
+  def assertRecorded(self, test_id, expected, retries):
+    self.test_lib.assertIn(test_id, self.last_execution_times)
+    self.test_lib.assertListEqual(expected['last_execution_time'][:retries],
+                                  self.last_execution_times[test_id])
+
+
+class TestResultsMock(object):
+  def __init__(self, test_lib):
+    self.results = []
+    self.test_lib = test_lib
+
+  def log(self, test_name, runtime_ms, actual_result):
+    self.results.append((test_name, runtime_ms, actual_result))
+
+  def assertRecorded(self, test_id, expected, retries):
+    test_results = [
+        (test_id[1], runtime_ms, 'PASS' if exit_code == 0 else 'FAIL')
+        for runtime_ms, exit_code in zip(expected['runtime_ms'][:retries],
+                                         expected['exit_code'][:retries])
+    ]
+    for test_result in test_results:
+      self.test_lib.assertIn(test_result, self.results)
+
+
+class TaskManagerMock(object):
+  def __init__(self):
+    self.running_groups = []
+    self.check_lock = threading.Lock()
+
+    self.had_running_parallel_groups = False
+    self.total_tasks_run = 0
+
+  def run_task(self, task):
+    test_group = task.test_name.split('.')[0]
+
+    with self.check_lock:
+      self.total_tasks_run += 1
+      if test_group in self.running_groups:
+        self.had_running_parallel_groups = True
+      self.running_groups.append(test_group)
+
+    # Delay as if real test were run.
+    time.sleep(0.001)
+
+    with self.check_lock:
+      self.running_groups.remove(test_group)
+
+
+class TaskMockFactory(object):
+  def __init__(self, test_data):
+    self.data = test_data
+    self.passed = []
+    self.failed = []
+
+  def get_task(self, test_id, execution_number=0):
+    task = TaskMock(test_id, execution_number, self.data[test_id])
+    if task.exit_code == 0:
+      self.passed.append(task)
+    else:
+      self.failed.append(task)
+    return task
+
+  def __call__(self, test_binary, test_name, test_command, execution_number,
+               last_execution_time, output_dir):
+    return self.get_task((test_binary, test_name), execution_number)
+
+
+class TaskMock(object):
+  def __init__(self, test_id, execution_number, test_data):
+    self.test_id = test_id
+    self.execution_number = execution_number
+
+    self.runtime_ms = test_data['runtime_ms'][execution_number]
+    self.exit_code = test_data['exit_code'][execution_number]
+    self.last_execution_time = (
+        test_data['last_execution_time'][execution_number])
+    self.test_command = None
+    self.output_dir = None
+
+    self.test_binary = test_id[0]
+    self.test_name = test_id[1]
+    self.task_id = (test_id[0], test_id[1], execution_number)
+
+  def run(self):
+    pass
+
+
+@contextlib.contextmanager
+def guard_temp_dir():
```

Comment:
This is not a mock. Either move this and below to another file (back to _tests) or rename this file something like `gtest_parallel_test_utils.py`. WDYT?

Reply:
I think I prefer moving these and below back to the top of the test file (gtest_parallel_mocks.py is a good name). Up to you though.